update mgt provider docs (#10645)
microsoftgraph/microsoft-graph-docs: concepts/toolkit/providers.md

# Microsoft Graph Toolkit providers

> The Microsoft Graph Toolkit providers enable your application to authenticate with and access Microsoft Graph to retreive the data to be rendered in the components. Each provider encapsulates and provides the implementation for acquiring the access token to call the Microsoft Graph APIs, so that you don't have to write this code yourself. You can also use the providers on their own, without components, to quickly implement authentication for your app and make your own calls to Microsoft Graph via the JavaScript SDK client.

Reviewer: ```suggestion
The Microsoft Graph Toolkit providers enable your application to authenticate with Microsoft Identity and access Microsoft Graph in only few lines of code. Each provider handles user authentication and acquiring the access tokens to call Microsoft Graph APIs, so that you don't have to write this code yourself. 

You can use the providers on their own, without components, to quickly implement authentication for your app and make calls to Microsoft Graph via the JavaScript client SDK.

The providers are required when using the Microsoft Graph Toolkit components as the components use them to access Microsoft Graph. If you already have your own authentication and simply want to use the components, you can use a [custom provider](./providers/custom.md) instead.
```

The Toolkit includes the following providers:

|Providers|Description|
|---------|-----------|

> |[Msal](./providers/msal.md)|Uses MSAL.js to sign in users and acquire tokens to use with Microsoft Graph.|

Reviewer: ```suggestion
|[Msal](./providers/msal.md)|Uses MSAL.js to sign in users and acquire tokens to use with Microsoft Graph in a web application.|
```

|[SharePoint](./providers/sharepoint.md)|Authenticates and provides Microsoft Graph access to components inside of SharePoint web parts.|
|[Teams](./providers/teams.md)|Authenticates and provides Microsoft Graph access to components inside of Microsoft Teams tabs.|
|[Proxy](./providers/proxy.md)|Allows the use of backend authentication by routing all calls to Microsoft Graph through your backend.|
|[Custom](./providers/custom.md)|Create a custom provider to enable authentication and access to Microsoft Graph with your application's existing authentication code.|

## Initializing a provider

To use a provider in your app, you need to initialize a new provider and then set it as the global provider in the Providers namespace. We recommend doing this before you start using any of the components. You can do this one of two ways:

**Option 1: Initialize in code**

To initalize a provider in your JavaScript code, create a new provider instance and set the value of the `Providers.globalProvider` property to the provider you'd like to use. The following example shows how to use the MsalProvider.

```js

> Providers.globalProvider = new MsalProvider({

Reviewer: ```suggestion
import {Providers, MsalProvider } from "@microsoft/mgt";
Providers.globalProvider = new MsalProvider({
```

>   clientId: '[YOUR_CLIENT_ID]'

Reviewer: ```suggestion
  clientId: 'YOUR_CLIENT_ID'
```

});
```

**Option 2: Use the provider component**

> You can also use the component version of the provider directly in your HTML. Behind the scenees, a new provider is initialized and set as the global provider. The following example shows how to use the MsalProvider.

Reviewer: ```suggestion
You can also use the component version of the provider directly in your HTML. Behind the scenes, a new provider is initialized and set as the global provider. The following example shows how to use the MsalProvider.
```

```HTML
<script src="https://unpkg.com/@microsoft/mgt/dist/bundle/mgt-loader.js"></script>

> <mgt-msal-provider client-id="[YOUR_CLIENT_ID]"></mgt-msal-provider>

Reviewer: ```suggestion
<mgt-msal-provider client-id="YOUR_CLIENT_ID"></mgt-msal-provider>
```

});
```

**Option 2: Use the provider component**

You can also use the component version of the provider directly in your HTML. Behind the scenes, a new provider is initialized and set as the global provider. The following example shows how to use the MsalProvider.

```HTML
<script src="https://unpkg.com/@microsoft/mgt/dist/bundle/mgt-loader.js"></script>
<mgt-msal-provider client-id="YOUR_CLIENT_ID"></mgt-msal-provider>
```

## Provider state

The provider keeps track of the user's authentication state and communicates this to the components. For example, when a user successfully signs in, the `ProviderState` is updated to `SignedIn`, signaling to the components that they are now able to make calls to Microsoft Graph. The `ProviderState` enum defines three states, as shown.

```ts
export enum ProviderState {
  Loading,
  SignedOut,
  SignedIn
}
```

In some scenarios, you will want to show certain functionality or perform an action only after a user has successfully signed in. You can access and check the provider state, as shown in the following example.

```js
import { Providers, ProviderState } from '@microsoft/mgt'

//assuming a provider has already been initialized

> if (Providers.globalProvider.state == ProviderState.SignedIn) {

Reviewer: ```suggestion
if (Providers.globalProvider.state === ProviderState.SignedIn) {
```

  //your code here
}
```

## Getting an access token

Each provider exposes a function called `getAccessToken` that can retreive the current access token or retrieve a new access token for the provided scopes. The following example shows how to get a new access token with the `User.Read` permission scope.

```js
import { Providers, ProviderState } from "@microsoft/mgt";

//assuming a provider has already been initialized

> if (Providers.globalProvider.state == ProviderState.SignedIn) {

Reviewer: ```suggestion
if (Providers.globalProvider.state === ProviderState.SignedIn) {
```

The `callbackFunction` function will be called when a provider is changed or when the state of a provider changes. A `ProvidersChangedState` enum value will be passed to the function to indicate what updated.

> - `getAccessToken({'scopes': scopes[]}) : Promise<string>`

Reviewer: `getAccessToken` is a method on the provider itself, not the Providers namespace. What do you think about adding a section above the 'Providers namespace' section:

## Getting an access token

Each provider exposes a function called `getAccessToken` that can retreive the current access token or retreive a new access token for provided scopes. For example

```tsx
import {Providers} from "@microsoft/mgt";

// assuming a provider has been initialized prior to this

if (Providers.globalProvider.state === ProviderState.SignedIn) {
  const token = Provider.globalProvider.getAccessToken({scopes: ['User.Read']})
}
```

The `getAccessToken` method accepts one optional parameter with a list of scopes. If a provider does not support incremental consent, the parameter will not have an effect.

Reviewer: Also consider adding a section prior to this on checking whether the provider is signed in, and move the section on calling the graph above as well.

Author: Added this section and rearranged the other ones

}
```

> ## Providers namespace

Reviewer: I really like this flow of all the different sections with examples. I think we can remove this entire `Providers namespace` section and replace it with a section on getting notified when the provider state changed via the `onProviderUpdated` function. It can go right under the `Provider State` section.

Setting the globalProvider is already covered in the above sections so I think it can be safely removed.

Author: Do you have an example of `onProviderUpdated` I can add?

Author: I added 2 more sections - user login/logout and permission scopes

Reviewer: Yeah, I think this might be a good example:

```ts
import { Providers, ProviderState } from "@microsoft/mgt";

const providerStateChanged = () => {
  if (Providers.globalProvider.state === ProviderState.SignedIn) {
    // user is now signed in
  }
}

// register a callback for when the state changes
Providers.onProviderUpdated(providerStateChanged);

// remove callback if necessary
Providers.removeProviderUpdatedListener(providerStateChanged);
```

    Providers.globalProvider = new MsalProvider(msalConfig)
}
```
## User Login/Logout

Once, you have the right providers initialized for your application, you can add the Toolkit's [Login component](./components/login.md) to easily and quickly implement user login and logout. The component works with the provider out of the box to handle all of the authentication logic and login/logout functionality. The following is an example of using the MsalProvider and the Login component.

```HTML
<script src="https://unpkg.com/@microsoft/mgt/dist/bundle/mgt-loader.js"></script>
<mgt-msal-provider client-id="YOUR_CLIENT_ID"></mgt-msal-provider>
<mgt-login></mgt-login>
```

>  In scenarios where you want to implement this yourself, rather than using the Toolkit's Login component, you can do so by using the `login` and `logout` methods.

Reviewer: ```suggestion
 In scenarios where you want to implement this yourself, rather than using the Toolkit's Login component, you can do so by using the `login` and `logout` methods of the provider.
```
